Add an ADR about secret key usage. (#26146)
edx/edx-platform: docs/decisions/0008-secret-key-usage.rst

Django Secret Key Usage
-----------------------

Status
======

Accepted

Context
=======

Django's Secret key is used by django to cryptographically sign various
capabilities in django.  It is also used by the edx-platform in various cases
as a `pepper`_ when hashing data to anonymize it while keeping it consistent.
Some of the uses of the ``SECRET_KEY`` in edx-platform were not resilient to
the key being rotated.  As a part of ARCHBOM-1646 those code paths are being
updated so that regular rotation of the ``SECRET_KEY`` will cause temporary
issues but will not fully break things.  ie. Sessions might get invalidated or
data might not be corelatable over the change boundry but the change is
otherwise safe.

.. _pepper: https://en.wikipedia.org/wiki/Pepper_(cryptography)

Decisions
=========

1. Any uses of ``SECRET_KEY`` should be resilient to that key being rotated.

> 2. Wherever the secret key is used, it should also document the impact of
>    rotating the key.

Reviewer: ```suggestion
2. Wherever the secret key is used, it should also document the impact of
   rotating the key or of exposure of the key to an attacker.
```

I'm not sure about the exact phrasing, but I'd like to request that people document the consequences of exposure as well.

Author: What do you think about just making that a separate  item.
```
3. Wherever the secret key is used, it should also document the consequences of the key being exposed.
```

Reviewer: Yeah, I like that.

Reviewer: Hmm, but also not just impact of rotating the key -- any process that's required, too.

Reviewer: (I mean, really, we should have these for all secrets... separate ADR?)

Reviewer: (I guess it would be an OEP, really.)

Django Secret Key Usage
-----------------------

Status
======

> Accepted

Author: ```suggestion
Pending
```

Context
=======

Django's Secret key is used by django to cryptographically sign various
capabilities in django.  It is also used by the edx-platform in various cases
as a `pepper`_ when hashing data to anonymize it while keeping it consistent.
Some of the uses of the ``SECRET_KEY`` in edx-platform were not resilient to
the key being rotated.  As a part of ARCHBOM-1646 those code paths are being
updated so that regular rotation of the ``SECRET_KEY`` will cause temporary
issues but will not fully break things.  ie. Sessions might get invalidated or

> data might not be corelatable over the change boundry but the change is

Reviewer 2: spelling: correlatable

> data might not be corelatable over the change boundry but the change is

Author: ```suggestion
data might not be correlatable over the change boundary but the change is
```

otherwise safe.

.. _pepper: https://en.wikipedia.org/wiki/Pepper_(cryptography)

Decisions
=========

1. Any uses of ``SECRET_KEY`` should be resilient to that key being rotated.

2. Wherever the ``SECRET_KEY`` is used, it should also document the impact of
   rotating the key.

> For example:

Reviewer: rST nit: This ends up styled as a heading while the other "for example" does not. Maybe something to do with line breaks and colons.

    Rotation is a one-time discontinuity in tracking metrics and should be
    accompanied by a heads-up to data researchers

3. Whenever the ``SECRET_KEY`` is used, it should also document the
   consequences of the key being exposed.

For example:

    Exposure of secret could result in identifying the tracking data for users
    if their actual session keys are also known

Consequences
============

> - The ``SECRET_KEY`` should only be used in cases, where rotation can be done

Reviewer: ```suggestion
- The ``SECRET_KEY`` should only be used in cases where rotation can be done
```
syntax nit

  quickly.  Because the secret is shared, not changing it quickly can expose

>   multiple features to attack if any one usage is more expensive than others.

Reviewer: ```suggestion
  multiple features to attack if any one usage is more expensive to rotate than others.
```
suggested clarification

  If you are considering using the ``SECRET_KEY`` in a situation where the act

>   of rotation will be expensive(time, coordination, preparation) you should

Reviewer: ```suggestion
  of rotation will be expensive (time, coordination, preparation) you should
```
nit: missing space
